File: ckl-dotnet/ckl-dotnet/ModuleLoader.cs
using System;
using System.IO;

namespace CheckerLang
{
    public class ModuleLoader
    {
        public static string LoadModule(string moduleidentifier, Environment environment, SourcePos pos) {
            try 
            {
                var assembly = typeof(Environment).Assembly;
                var strm = assembly.GetManifestResourceStream("checkerlang.module-" + moduleidentifier.ToLower());
                if (strm != null)
                {
                    return new StreamReader(strm).ReadToEnd();
                }

                var reader = GetModuleStream(moduleidentifier, environment, pos);
                if (reader != null) 
                {
                    return reader.ReadToEnd();
                }
                    
                throw new ControlErrorException(new ValueString("ERROR"), "Module " + new FileInfo(moduleidentifier).Name + " not found", pos);
            } 
            catch (IOException) 
            {
                throw new ControlErrorException(new ValueString("ERROR"), "Module " + new FileInfo(moduleidentifier).Name + " not found", pos);
            }
        }

        private static StreamReader GetModuleStream(String moduleidentifier, Environment environment, SourcePos pos) 
        {
            string userdir = System.Environment.GetFolderPath(System.Environment.SpecialFolder.UserProfile);
            string moduledir = userdir + "/.ckl/modules";
            string module = moduledir + "/" + new FileInfo(moduleidentifier).Name;
            if (File.Exists(module)) return new StreamReader(module);
            if (environment.IsDefined("checkerlang_module_path")) 
            {
                var modulepath = environment.Get("checkerlang_module_path", pos);
                foreach (var dir in modulepath.AsList().GetValue()) 
                {
                    module = dir.AsString().GetValue() + "/" + new FileInfo(moduleidentifier).Name;
                    if (File.Exists(module)) return new StreamReader(module);
                }
            }
            module = "./" + new FileInfo(moduleidentifier).Name;
            if (File.Exists(module)) return new StreamReader(module);
            return null;
        }
    }
}

File: ckl-java/src/main/ch/checkerlang/ModuleLoader.java
package ch.checkerlang;

import java.io.*;
import java.nio.charset.StandardCharsets;

import ch.checkerlang.values.Value;
import ch.checkerlang.values.ValueList;

public class ModuleLoader {
    public static String loadModule(String moduleidentifier, Environment environment, SourcePos pos) {
        try {
            InputStream strm = ModuleLoader.class.getResourceAsStream("/module-" + moduleidentifier.toLowerCase());
            if (strm == null) strm = getModuleStream(moduleidentifier, environment, pos);
            if (strm == null) throw new ControlErrorException("Module " + new File(moduleidentifier).getName() + " not found", pos);
            BufferedReader rdr = new BufferedReader(new InputStreamReader(strm, StandardCharsets.UTF_8));
            StringBuilder result = new StringBuilder();
            try {
                String line = rdr.readLine();
                while (line != null) {
                    result.append(line).append("\n");
                    line = rdr.readLine();
                }
            } finally {
                rdr.close();
            }
            return result.toString();
        } catch (IOException e) {
            throw new ControlErrorException("Module " + new File(moduleidentifier).getName() + " not found", pos);
        }
    }

    private static InputStream getModuleStream(String moduleidentifier, Environment environment, SourcePos pos) throws IOException {
        File moduledir = new File(new File(System.getProperty("user.home"), ".ckl"), "modules");
        File module = new File(moduledir, new File(moduleidentifier).getName());
        if (module.exists()) return new FileInputStream(module);
        if (environment.isDefined("checkerlang_module_path")) {
            ValueList modulepath = environment.get("checkerlang_module_path", pos).asList();
            for (Value dir : modulepath.getValue()) {
                module = new File(dir.asString().getValue(), new File(moduleidentifier).getName());
                if (module.exists()) return new FileInputStream(module);
            }
        }
        module = new File(".", new File(moduleidentifier).getName());
        if (module.exists()) return new FileInputStream(module);
        return null;
    }
}
